File: src/main/java/offer/Offer56_2.java
package offer;

/**
 * @author jingxinwu
 * @date 2021-11-25 12:51 上午
 */
public class Offer56_2 {


    //todo
    public int singleNumber(int[] nums) {
        int ones = 0, twos = 0;
        for (int num : nums) {
            ones = ones ^ num & ~twos;
            twos = twos ^ num & ~ones;
        }
        return ones;
    }

}

File: src/main/go/leetcode/lists/offer/Offer56_2.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/5 16:11
 */

func singleNumber(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	ones, twos := 0, 0
	for _, num := range nums {
		ones = ones ^ num&(^twos)
		twos = twos ^ num&(^ones)
	}
	return ones
}
